File: backend/src/main/java/incident/management/system/service/AnalyticsService.java
package incident.management.system.service;

import incident.management.system.dto.analytics.HeatmapResponse;
import incident.management.system.dto.analytics.ParetoResponse;
import incident.management.system.dto.analytics.RepeatSignalResponse;
import incident.management.system.dto.analytics.VolumeSpeedResponse;
import incident.management.system.dto.analytics.WorkloadEntry;

import java.time.LocalDate;
import java.util.List;

/**
 * Analytical & quality-engineering queries backing the /analytics page.
 *
 * <p>All metrics are computed on demand from the {@code incidents} table with
 * database-level time bucketing ({@code DATE_TRUNC}) and SQL window functions —
 * the service never pulls raw unbounded datasets to aggregate in memory.
 */
public interface AnalyticsService {

    /**
     * Time-bucketed volume/speed analytics over {@code [start, end]}.
     *
     * @param compare when true, also computes period-over-period deltas against
     *                the previous window of identical length.
     */
    VolumeSpeedResponse getVolumeSpeed(LocalDate start, LocalDate end,
                                       Long departmentId, boolean compare);

    /** Category Pareto (80/20) analysis over the window. */
    ParetoResponse getPareto(LocalDate start, LocalDate end, Long departmentId);

    /** Hour-of-day × day-of-week incident density matrix over the window. */
    HeatmapResponse getHeatmap(LocalDate start, LocalDate end, Long departmentId);

    /** Rule-based repeat-incident signals (≥ 3 same station+category in 14 days). */
    RepeatSignalResponse getRepeatSignals(LocalDate start, LocalDate end, Long departmentId);

    /** ADMIN-scoped aggregate team workload per evaluator. */
    List<WorkloadEntry> getWorkload(LocalDate start, LocalDate end, Long departmentId);
}

File: frontend/services/analyticsService.ts
import apiClient from '@/lib/api-client';
import type {
  HeatmapResponse,
  ParetoResponse,
  RepeatSignalResponse,
  VolumeSpeedResponse,
  WorkloadEntry,
} from '@/types/analytics';

/** Shared analytics query parameters — the global control bar of /analytics. */
export interface AnalyticsQueryParams {
  /** ISO date (yyyy-MM-dd). */
  startDate: string;
  /** ISO date (yyyy-MM-dd), inclusive. */
  endDate: string;
  /** Optional department filter. */
  departmentId?: number;
}

/** Time-bucketed volume, resolution quality, MTTR, time-to-claim + deltas. */
export async function getVolumeSpeed(
  params: AnalyticsQueryParams,
  compare: boolean,
): Promise<VolumeSpeedResponse> {
  const { data } = await apiClient.get<VolumeSpeedResponse>('/api/analytics/volume-speed', {
    params: { ...params, compare },
  });
  return data;
}

/** Category Pareto (80/20) analysis — server-side cumulative percentages. */
export async function getPareto(params: AnalyticsQueryParams): Promise<ParetoResponse> {
  const { data } = await apiClient.get<ParetoResponse>('/api/analytics/pareto', { params });
  return data;
}

/** Hour-of-day × day-of-week shift heatmap. */
export async function getHeatmap(params: AnalyticsQueryParams): Promise<HeatmapResponse> {
  const { data } = await apiClient.get<HeatmapResponse>('/api/analytics/heatmap', { params });
  return data;
}

/** Rule-based repeat-incident signals. */
export async function getRepeatSignals(params: AnalyticsQueryParams): Promise<RepeatSignalResponse> {
  const { data } = await apiClient.get<RepeatSignalResponse>('/api/analytics/repeat-signals', { params });
  return data;
}

/** ADMIN-scoped team workload distribution. */
export async function getWorkload(params: AnalyticsQueryParams): Promise<WorkloadEntry[]> {
  const { data } = await apiClient.get<WorkloadEntry[]>('/api/analytics/workload', { params });
  return data;
}
